### PR TITLE
Groupby reborn

### PR DESCRIPTION
**Pull request by chaupow**
anyone wanna review?

this PR adds a groupby node and includes the multiway merge algorithm from stxxl.

### CODE REVIEW

File: thrill/api/groupby.hpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            // if vector is full save to disk
+            if (incoming.size() == FIXED_VECTOR_SIZE) {
+                FlushVectorToFile(incoming);
+                incoming.clear();
+            }
+            // store incoming element
+            const auto elem = reader.template Next<ValueIn>();
+            incoming.push_back(elem);
+            // LOG << "received " << elem << " with key " << key_extractor_(incoming.back());
+            ++totalsize;
```
Comment (emiswelt) [inline thrill/api/groupby.hpp]: Is counting all incoming elements manually really necessary?

Comment (chaupow) [inline thrill/api/groupby.hpp]: @emiswelt good catch.

**Comment by chaupow** [comment]
please do not merge this PR yet. it has some bugssss

**Comment by emiswelt** [comment]
I approve.

**Comment by bingmann** [comment]
Sorry, this was completely premature!
The files core/stxxl_... contain a *different namespace*.
That must be fixed, they dont uncrustify correctly!

**Comment by bingmann** [comment]
They also contain superfluous classes like noncopyable. -> change to c++11 non-copyableness.

**Comment by bingmann** [comment]
The cmake contains a include_directory to a non-existing dir.
Your copyright is missing on the files.
Remove my GPL copyright! You included a GPL file in our project.That makes the whole project GPL! This is only not a fatal error because the file is from me.
Use round up to power of two that already exists in common.
think about renaming much of the STXXL_ macros to THRILL_
write in the header where you got the code from, who made it, and under which license you are using it.

**Comment by bingmann** [comment]
I have reverted the merge by an evil force push, because it is really not ready.
Fix it, asap.

**Comment by bingmann** [comment]
And then first READ the diff to the master for things that should not be.
https://github.com/thrill/thrill/compare/groupby_reborn#files_bucket

**Comment by emiswelt** [comment]
Ok, I'll stop reviewing then. 

There's a lot of stuff in there I did not even think about (like GPL).

**Comment by chaupow** [comment]
sorry. had my head in the clouds and.. didn't think of license at all.


> The files core/stxxl_... contain a different namespace.
> That must be fixed, they dont uncrustify correctly!

- [ ] todo

> They also contain superfluous classes like noncopyable. -> change to c++11 non-copyableness.

- [ ] todo

> The cmake contains a include_directory to a non-existing dir.

- [ ] todo

> Your copyright is missing on the files.

- [ ] todo

> Remove my GPL copyright! You included a GPL file in our project.

- [ ] todo

> Use round up to power of two that already exists in common.

- [ ] todo

> think about renaming much of the STXXL_ macros to THRILL_

- [ ] todo

> write in the header where you got the code from, who made it, and under which license you are using it.

- [ ] todo

> And then first READ the diff to the master for things that should not be.

- [ ] todo

**Pull request merged**